[HttpKernel] ConfigDataCollector to return known data without the need of a Kernel

topikito opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #40534
| License       | MIT
| Doc PR        | 

Sets `$this->data` with information that can be known without the need of a `Kernel`.

topikito committed:
[HttpKernel] ConfigDataCollector to return known data without the need of a Kernel

derrabus commented:
How is it possible that we get `null` here in the first place? Shouldn't we always be able to determine that piece of information?

ciaranmcnulty commented:
@derrabus Not when this condition is false https://github.com/symfony/symfony/blob/b5693be406dfeed8b768a4a9af997338e3e72450/src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php#L81-L93

fabpot commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -175,9 +175,9 @@ public function getSymfonyMinorVersion()
+    public function isSymfonyLts(): ?bool
     {
-        return $this->data['symfony_lts'];
+        return $this->data['symfony_lts'] ?? null;
```
I would return `false` instead of `null`

topikito replied:
After having reviewed again the code, could there be a case to be made around leaving the method's signature as it is, and moving the assignment out of the `if` ?

Before the `if`, on line _47_, there is already a use of a constant of  the class `Kernel` (Plus, `Kernel` belongs to the same component as this class, so it will always exist):
```php
            'symfony_version' => Kernel::VERSION,
```

 There are some assignments inside the if which use `Kernel` as well, not the attribute `$kernel`.

```php
            $this->data['symfony_state'] = $this->determineSymfonyState();
            $this->data['symfony_minor_version'] = sprintf('%s.%s', Kernel::MAJOR_VERSION, Kernel::MINOR_VERSION);
            $this->data['symfony_lts'] = 4 === Kernel::MINOR_VERSION;
            $eom = \DateTime::createFromFormat('d/m/Y', '01/'.Kernel::END_OF_MAINTENANCE);
            $eol = \DateTime::createFromFormat('d/m/Y', '01/'.Kernel::END_OF_LIFE);
            $this->data['symfony_eom'] = $eom->format('F Y');
            $this->data['symfony_eol'] = $eol->format('F Y');
```

I would suggest bringing these assignments out of the `if`, as they do not depend on the optional attribute `$kernel` to exist. They depend on the `Kernel` class, which will always exist because it's in the same component.

nicolas-grekas replied:
I'm fine with moving these outside of the "if" block.

topikito replied:
Changes made

ro0NL commented:
we should only rely on the `$this->kernel` instance for bundle aggregration IMHO.

then always get remaining metadata from static `Kernel` symbol (which is a hard dep)

topikito commented:
I have suggested a different solution as to the suggestion of @fabpot to return `false`: bring out of the `if` block all assignments that rely on `Kernel` and not on `$kernel`.

https://github.com/symfony/symfony/pull/40535#discussion_r598559548

topikito commented:
I have extracted the corresponding bits outside of the `if` block.

For the tests, I decided to add extra checks for all the content I extracted. I required a way to generate the same output for `determineSymfonyState()` so I decided to copy the behaviour. The reason behind this is that I assume it's not desired to rewrite the tests for each version, so it should rather expect to have the same logic in the tests. If the logic in the class would change, the test should catch it. If this is not acceptable, please let me know how to proceed to test that particular getter.

stof commented on src/Symfony/Component/HttpKernel/Tests/DataCollector/ConfigDataCollectorTest.php:
```diff
@@ -58,6 +65,53 @@ public function testLegacy()
+        $this->assertSame($eol, $c->getSymfonyEol());
+    }
+
+    private function determineSymfonyState(): string
```
copy-pasting the logic does not provide any guarantee of correctness. What the test is verifying right now is that you were able to do a copy-paste without mistake. But it makes it harder to change the logic, as we would need to redo the copy-paste.

I would instead change the assertion, to assert only that it is one of the allowed values (without caring about which one to avoid having to update tests when reaching the eol date indeed)

topikito replied:
Thanks for the feedback @stof . I will change it now.

Just to educate myself, how does this compare to the check for `4 === Kernel::MINOR_VERSION` that was currently in place for the value `symfony_lts`? Isn't it the same type of procedure?

stof replied:
well, the `4 === Kernel::MINOR_VERSION` condition does not have lots of copy-pasted logic in it, and is unlikely to ever be modified (at least until we change the release schedule entirely). But it is true that it does not really guarantee the correctness either.

topikito replied:
Thanks for clarifying - Changes have been submitted

nicolas-grekas commented:
The descrition+title of the PR need an update now ;)

Nyholm commented:
Awesome. Congratulations to your first contribution!
